### ISSUE TITLE
root credentials from logrecordd

### ISSUE THREAD

**Issue by JayFoxRox** [najashark/tplink-tapo-c225-re#1]
Hi,

> Now we need to find the root credentials because we are not lucky enough to find the password in the source code.

logrecordd connects to telnet if diagnose mode is turned on.

How the password is retrieved changed a lot since older versions (some stored plaintext root:slprealtek for example).
I have analyzed a bunch of relatively recent firmwares and some firmware for the C310 does something interesting:

User: "root"
Password: 
```
pcVar4 = (char *)rsa_decrypt("EctjjP4YY0W4OYIULZppwTpsQAG1twYBCq1pyELaC89vn04L0V5fRdTV5rua OjaNTmFvQUDUTa/QJ5vt/sUOq/8r3qaLCy68QUZNZJCbm5Oy+fXHDonGovTHgij1xBLPhpX6kmR+iVp1Tn1qe6Ioyq Aij2koWb46T59Y+DoSFbk="
                                       ,0xac);
```

with `rsa_decrypt` from libdecrypter.so:

```
void rsa_decrypt(int param_1,int param_2) {
  char acStack_340 [816];
  
  memcpy(acStack_340,
         "MIICXAIBAAKBgQC4D6i0oD/Ga5qb//RfSe8MrPVIrMIGecCxkcGWGj9kxxk74qQNq8XUuXoy2PczQ30BpiRHrlkbtB EPeWLpq85tfubTUjhBz1NPNvWrC88uaYVGvzNpgzZOqDC35961uPTuvdUa8vztcUQjEZy16WbmetRjURFIiWJgFCmem yYVbQIDAQABAoGAZ7lLVR7JUcPpyOegiuJbOEVvpJjWblfGY0rEURZRizU33yuFT77xKUOsvWLPS7BIjdlWsJ5r0NTU mGfLees71DqxfedQb3kSamBMErHu/jXeRi7MaGCLTyO9Ae+0+PFAHUdld1QjX50mAZD/+fbDPv7zUebNrMkFzbl99ct YznECQQDwiHNBP51VOF2f8pRmncr5QpZmWZb3ZNPw9Qc/97kn7Xs8zGz1NBEUA9SLcF25KXv6GpVdr6X7BuhXeB9HZp NTAkEAw+WYCnB1KjYz68FiIYkA63lrvnN2IG9pllkt+az6XFdXgOdqodoi3pWvmhkU55Z2N0okuGjm93qO1pebC3JcP wJAOsn+8Yms2LFoILnXj6UtgPLHc8id32Wjb5dT6EyR0rJ2louYbe4F5pBxGIukPKdpB94Ld9SAivRLQWW4r2jgxQJB AJo1+D1nj+Rd7PuPLXfmyRGVcQrpC7m22vDfXUDqOeBNZXX1Ns0Y0lBUl3sAeaNhn8ggls2QzxlMonstt4EIUrMCQCg OhLJ1SoUw1RE+b3x5hpuibFpiknX9MAjNxMJ1vHQFgsYTIVlD9LdHc/+nGbeXhzi0BJ2h3R5tmQmIseOs9f0="
         ,0x32d);
  strlen(acStack_340);
  private_decrypt(param_1,param_2,acStack_340);
  return;
}
```

`private_decrypt` starts out by wrapping the third argument in `"-----BEGIN RSA PRIVATE KEY-----"` and `"-----END RSA PRIVATE KEY-----"`.
The rest should be pretty straight-forward.

I haven't actually tried decrypting this myself yet - but it might be what you need.
If it's not what you need, it's probably a variation of what we'll need.

**Comment by JayFoxRox** [comment]
That'd be `root:slplzgjipc` (untested)

**Comment by najashark** [comment]
@JayFoxRox I'm able to get the root password after fully dumping the SPI NAND flash
through some strings and grep `root:`. I'm able to get this hash

`root:$1$meUCOF2P$ZbDuDUPjhxSekjJNzj9ol/:0:0:root:/root:/bin/ash`

and I able to crack it, which you can use `slpmstar` as `root` password, 
I think this password will be same for all tapo with sigmastar/mstar SoC

**Comment by DrmnSamoLiu** [comment]
It seems you guys are both very capable of hardware and software reverse engineering, I believe finding the firmware decryption key is only a matter of time now. Wish you all best of luck :)
@JayFoxRox Since @najashark shared the mtd, I think I don't need to go out and buy a new tapo cam for dumping now. (As I mentioned in one of the issues you opened.)

**Comment by JayFoxRox** [comment]
> I added https://github.com/najashark/tplink-tapo-c225-re/blob/main/firmware.zip from firmware mtd partitions

Thanks a lot @najashark !

Unfortunately, that dump doesn't seem to contain the most interesting binaries (uhttpd, dsd, cet, cloud-iot, logrecordd, libdecrypter.so) 🤔 

In some recent firmwares (example: Tapo_C210v1_en_1.3.0_Build_220830_Rel.69909n_u_1663826839743.bin) those are in a separate partition; probably so they can update the tplink stuff independently from the rest of the OS:

```bash
% binwalk Tapo_C210v1_en_1.3.0_Build_220830_Rel.69909n_u_1663826839743.bin
DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
2048          0x800           uImage header, header size: 64 bytes, header CRC: 0x7C640FC2, created: 2022-08-30 11:07:59, image size: 67528 bytes, Data Address: 0x0, Entry Point: 0x0, data CRC: 0xE32CC573, OS: Firmware, CPU: ARM, image type: OS Kernel Image, compression type: lzma, image name: "MVX4##I6B0gd1dfee1CM_UBT1501#XVM"
2112          0x840           xz compressed data
133632        0x20A00         uImage header, header size: 64 bytes, header CRC: 0x1173ABCB, created: 2022-08-30 11:25:04, image size: 1465296 bytes, Data Address: 0x20008000, Entry Point: 0x20008000, data CRC: 0xB8B9EE83, OS: Linux, CPU: ARM, image type: OS Kernel Image, compression type: none, image name: "Linux-4.9.84"
148579        0x24463         xz compressed data
148913        0x245B1         xz compressed data
1600000       0x186A00        Squashfs filesystem, little endian, version 4.0, compression:xz, size: 2347058 bytes, 667 inodes, blocksize: 131072 bytes, created: 2022-08-30 11:25:24
4065280       0x3E0800        Squashfs filesystem, little endian, version 4.0, compression:xz, size: 3771898 bytes, 253 inodes, blocksize: 131072 bytes, created: 2022-08-30 11:25:27
7998720       0x7A0D00        gzip compressed data, from Unix, last modified: 2022-08-30 11:25:20
8028628       0x7A81D4        gzip compressed data, from Unix, last modified: 2022-08-30 11:25:20
```

Contrast to your image:

```bash
% binwalk firmware.bin
DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
1048576       0x100000        uImage header, header size: 64 bytes, header CRC: 0xFBCEEB51, created: 2022-09-22 13:12:21, image size: 1697304 bytes, Data Address: 0x20008000, Entry Point: 0x20008000, data CRC: 0xDC49E7BF, OS: Linux, CPU: ARM, image type: OS Kernel Image, compression type: none, image name: "Linux-4.9.84"
1063555       0x103A83        xz compressed data
1063889       0x103BD1        xz compressed data
9437184       0x900000        Squashfs filesystem, little endian, version 4.0, compression:xz, size: 14734078 bytes, 599 inodes, blocksize: 65536 bytes, created: 2022-09-22 13:12:46
```

(which matches just the middle portion of the full flash, but unfortunately without the second squashfs)

So it looks like the partition after this one should also be a squashfs which has the interesting bits.
I'm not sure about the partition name - binwalk should let you know if you find the second squashfs (unless they encrypt it now)

Also, what firmware version are you running? The squashfs creation date seems to be from much earlier than the creation of the earliest C225 firmware.

**Comment by najashark** [comment]
That's correct, from my quick view. All of it just run from `/bin/main` while `/bin/monitor` is used to monitor the `/bin/main` if it crashes and restart it. here are the list of running process on my c225
![image](https://github.com/najashark/tplink-tapo-c225-re/assets/1461279/081041c5-3834-4cca-8c30-6175ef34e251)

**Comment by najashark** [comment]
I have dumped all the [mtdblock ](https://github.com/najashark/tplink-tapo-c225-re/tree/main/mtdblock) from firmware version 1.0.27_Build_231107_Rel.69776n except config.bin for obvious reason, but you can decrypt config.bin with this cyberchef recipe, make sure the header look like this. 

If the key is different, it can be found at `/tmp/base-files/etc/encrypt_key`

![image](https://github.com/najashark/tplink-tapo-c225-re/assets/1461279/83bc2085-7570-4053-b9c5-3514098a1be0)

```
https://gchq.github.io/CyberChef/#recipe=To_Hex('Space',0)Find_/_Replace(%7B'option':'Simple%20string','string':'4E%2056%204D%2050%202D%2043%204F%204E%2046%2049%2047%2000%2000%2000%2000%2001%20BC%2000%2000%2000%2020%201F%2000%2000%2044%2013%2001%2000%20'%7D,'',true,true,true,false)DES_Decrypt(%7B'option':'Hex','string':'1157B26D1360D7F6'%7D,%7B'option':'Hex','string':'000000000000000'%7D,'CBC/NoPadding','Hex','Raw')Zlib_Inflate(0,0,'Adaptive',false,false)
```

**Comment by JayFoxRox** [comment]
Very awesome. Quality work @najashark

I was just starting to look at my config dumps again (after the old DES decryption didn't work for me).
I was about to ask you to dump `/tmp/base-files/etc/encrypt_key` for me and accidentally saw that I still had this issue opened (which also meant I didn't get a notification yesterday).
I was already wondering why the decryption code used `+0x1C`, but your screenshot show it nicely.

I dumped my config using the `download_conf` DS request / HTTP, but it doesn't have that header.
I originally tried to decrypt my dumps using the old method (which had a hardcoded key, but also DES), but it didn't work.
I've tried your encryption key / recipe now and it doesn't seem to work for me (or it's a different format?).
I'll look into it in the next days probably (specifically where that encrypt_key is coming from - I assume it's derived from the `devId`?).

**Comment by najashark** [comment]
During firmware upgrade/downgrade using sdcard, `main` will call `/tmp/sdcard/factory_script.sh`, so we can enable SSH through this method

#### Enable SSH

1. Format SD Card to FAT32
2. Copy all files in https://github.com/najashark/tplink-tapo-c225-re/tree/main/enablessh to SD card
3. Insert to the SD card slot on camera
4. Reboot device

If SSH still not enabled, try restarting camera using TAPO apps.

**Comment by JayFoxRox** [comment]
> During firmware upgrade/downgrade using sdcard, `main` will call `/tmp/sdcard/factory_script.sh`, so we can enable SSH through this method

Yes, I also spotted it and hinted at it here: https://github.com/nervous-inhuman/tplink-tapo-c200-re/discussions/6#discussioncomment-7834165
Here are some of my WIP notes regarding this:

```
- format as vfat and insert in camera
    - should show up as /dev/mmcblk0p1 and mount vfat to /tmp/sdcard
- touch /sdcard/factory_up_boot.bin
- put your script in /sdcard/factory_script.sh
    - should be ran with args: "fw_upgrade" "fail"
```

Note that I didn't try the SD card script yet, mostly because I want an OTA RCE and I'm too lazy to prepare an SD card; these notes are purely based on code.

As suggested here, this should also happen if the upgrade fails (at least that's my theory).
The whole thing gets triggered by `msg_send` events and msg handlers, so the code gets a bit hard to follow.
It's very possible this also happens for OTA upgrades.

It would be nice if we could craft at least 2 payloads ("factory_script.sh"):
- Enable SSH forever by flashing some file + also drop a simple script which can be used to undo the hack (in case future firmwares check for it, or if a user wants to return to stock); you have partialle done this already ❤️ 
- Temporarily enable SSH (which I'd prefer)
- (+ potentially more things like a script to fixup all networking so the camera is definitely cloud-less)

I also want to revisit looking at that code as my notes are somewhat incomplete.

I think you should also add a README.md in the "enablessh" directory about where `factory_up_boot.bin` comes from. I assume it's a stock C225 firmware?
Also where did you get that dropbear bin from (is it straight from some openwrt repo)?
I assume it would cause trouble on a MIPS based device.
